drivers: pwm: nrfx: adjust PWM driver to fast PWM120

mstasiaknordic opened the pull request:
Fast PWM120 instance works with 320MHz clock, thus `pwm_nrfx_get_cycles_per_sec` needs to be adjusted.
Also, it uses cachable RAM, thus sys_cache function needs to be added to flush the cached sequence.

nika-nordic commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -176,6 +171,10 @@ static int pwm_nrfx_set_cycles(const struct device *dev, uint32_t channel,
 	seq_values_ptr_get(dev)[channel] = PWM_NRFX_CH_VALUE(compare_value, inverted);
 
+#ifdef CONFIG_NRFX_PWM120
+	sys_cache_data_flush_range((void *)config->seq.values.p_raw, config->seq.length);
```
I think we need to check in runtime whether d-cache operation is needed or not, depending on `seq.values.p_raw` memory region attributes. You can take inspiration from this PR: https://github.com/zephyrproject-rtos/zephyr/pull/80015/files , see `UARTE_IS_CACHEABLE` .

Maybe in the future we could switch to `dmm` for this shim, but let's it keep simple as this is hotfix

masz-nordic commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -243,12 +242,11 @@ static int pwm_nrfx_set_cycles(const struct device *dev, uint32_t channel,
-	 * etc, in the pin set function. See issue #6958.
-	 */
+#ifdef CONFIG_NRFX_PWM120
+	*cycles = 320ul * 1000ul * 1000ul;
+#else
 	*cycles = 16ul * 1000ul * 1000ul;
-
+#endif
```
How about using this approach from UARTE:
https://github.com/zephyrproject-rtos/zephyr/commit/036a85b517deab873d60bf2eff552fb5e1a4afc9
https://github.com/zephyrproject-rtos/zephyr/commit/7c011741faecd60d482087c000666228064fb0af

mstasiaknordic replied:
Since it is a hotfix, could be postpone the changes until the DMM rework?

masz-nordic replied:
I don't see connection between this and DMM, but I'm fine with this staying temporarily if you post a follow-up PR.

gmarull commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -243,12 +249,11 @@ static int pwm_nrfx_set_cycles(const struct device *dev, uint32_t channel,
-	 * 16MHz from this function and handle the conversion with prescaler,
-	 * etc, in the pin set function. See issue #6958.
-	 */
+#ifdef CONFIG_NRFX_PWM120
+	*cycles = 320ul * 1000ul * 1000ul;
+#else
```
```suggestion
#ifdef CONFIG_NRFX_PWM120
	*cycles = 320ul * 1000ul * 1000ul;
#else
```

this won't work when enabling multiple instances.

mstasiaknordic commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -243,12 +249,17 @@ static int pwm_nrfx_set_cycles(const struct device *dev, uint32_t channel,
-	 * 16MHz from this function and handle the conversion with prescaler,
-	 * etc, in the pin set function. See issue #6958.
-	 */
+#if defined(NRF_PWM120)
+	const struct pwm_nrfx_config *config = dev->config;
+	if(config->pwm.p_reg == NRF_PWM120)
+	{
+		*cycles = 320ul * 1000ul * 1000ul;
+		
+		return 0;
+	}
+#endif
```
I am affraid that would require adding clock property to every nrf pwm node. If thats not an issue, then sure.

gmarull replied:
all that need it, yes. Just check the UART case.

gmarull replied:
it can be made optional, see

https://github.com/zephyrproject-rtos/zephyr/blob/124aae3876842154964726ba6e89565040648f9a/drivers/serial/uart_nrfx_uarte.c#L2290-L2294

mstasiaknordic committed:
dts: common: nordic: Add clock property for fast PWM120

Added hsfll120 clock for fast PWM120 nodes.

Signed-off-by: Michał Stasiak <michal.stasiak@nordicsemi.no>

mstasiaknordic committed:
drivers: pwm: nrfx: adjust PWM driver to fast PWM120

Fast PWM120 instance works with 320MHz clock, thus
pwm_nrfx_get_cycles_per_sec needs to be adjusted,
applying correct clock frequency.
Also, it uses cachable RAM, thus sys_cache function
needs to be added to flush the cached sequence.

Signed-off-by: Michał Stasiak <michal.stasiak@nordicsemi.no>

anangl commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -176,6 +182,12 @@ static int pwm_nrfx_set_cycles(const struct device *dev, uint32_t channel,
 
+#ifdef CONFIG_DCACHE
+	if (config->mem_attr & DT_MEM_CACHEABLE) {
+		sys_cache_data_flush_range((void *)config->seq.values.p_raw, config->seq.length);
```
```suggestion
		sys_cache_data_flush_range(seq_values_ptr_get(dev), config->seq.length);
```

anangl commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -333,13 +342,19 @@ static int pwm_nrfx_init(const struct device *dev)
 
+#define PWM_GET_MEM_ATTR(idx)								\
+	COND_CODE_1(PWM_HAS_PROP(idx, memory_regions),					\
+		(COND_CODE_1(DT_NODE_HAS_PROP(PWM_MEM_REGION(idx), zephyr_memory_attr), \
+			(DT_PROP(PWM_MEM_REGION(idx), zephyr_memory_attr)), (0))), (0))
```
```suggestion
		(DT_PROP_OR(PWM_MEM_REGION(idx), zephyr_memory_attr, 0)), (0))
```

anangl replied:
not addressed

mstasiaknordic replied:
Sorry, missed it. Corrected now

anangl commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -362,6 +377,11 @@ static int pwm_nrfx_init(const struct device *dev)
 		.pcfg = PINCTRL_DT_DEV_CONFIG_GET(PWM(idx)),		      \
+		.clock_freq = COND_CODE_1(DT_CLOCKS_HAS_IDX(PWM(idx), 0),     \
+			(DT_PROP(DT_CLOCKS_CTLR(PWM(idx)), clock_frequency)), \
+				(16ul * 1000ul * 1000ul)),		      \
```
```suggestion
			(16ul * 1000ul * 1000ul)),			      \
```

anangl commented on dts/common/nordic/nrf54h20.dtsi:
```diff
@@ -629,6 +629,7 @@
 				interrupts = <228 NRF_DEFAULT_IRQ_PRIORITY>;
 				power-domains = <&gpd NRF_GPD_FAST_ACTIVE1>;
 				#pwm-cells = <3>;
+				clocks = <&hsfll120>;
```
Could you place it right after `interrupts` for consistency with other peripherals?

anangl commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -243,11 +255,8 @@ static int pwm_nrfx_set_cycles(const struct device *dev, uint32_t channel,
-	 */
-	*cycles = 16ul * 1000ul * 1000ul;
+	const struct pwm_nrfx_config *config = dev->config;
+	*cycles = config->clock_freq;
```
Please add a blank line between these two. It's strange that checkpatch did not complain about that.

anangl commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -362,6 +378,11 @@ static int pwm_nrfx_init(const struct device *dev)
 		.seq.values.p_raw = pwm_##idx##_seq_values,		      \
 		.seq.length = NRF_PWM_CHANNEL_COUNT,			      \
 		.pcfg = PINCTRL_DT_DEV_CONFIG_GET(PWM(idx)),		      \
+		.clock_freq = COND_CODE_1(DT_CLOCKS_HAS_IDX(PWM(idx), 0),     \
+		(DT_PROP(DT_CLOCKS_CTLR(PWM(idx)), clock_frequency)),	      \
+			(16ul * 1000ul * 1000ul)),			      \
```
This indentation is misleading. It should be:
```suggestion
		.clock_freq = COND_CODE_1(DT_CLOCKS_HAS_IDX(PWM(idx), 0),     \
			(DT_PROP(DT_CLOCKS_CTLR(PWM(idx)), clock_frequency)), \
			(16ul * 1000ul * 1000ul)),			      \
```

anangl commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -333,13 +343,18 @@ static int pwm_nrfx_init(const struct device *dev)
+			PWM_MEM_REGION(idx)))))),			      \
 		())
 
+#define PWM_GET_MEM_ATTR(idx)								\
+	COND_CODE_1(PWM_HAS_PROP(idx, memory_regions),					\
```
```suggestion
#define PWM_GET_MEM_ATTR(idx)						      \
	COND_CODE_1(PWM_HAS_PROP(idx, memory_regions),			      \
```

mstasiaknordic committed:
drivers: pwm: nrfx: add pin retention

Add pin retenion if GPD is enabled for nRF54H20.

Signed-off-by: Michał Stasiak <michal.stasiak@nordicsemi.no>
